### PR TITLE
Usability improvements to failures in oc login

### PR DESCRIPTION
**Pull request by fabianofranz**
Use just TCP when trying to check if server is reachable. Don't default to 8443 if port was not provided, but still tries it as a fallback if 443 isn't reachable. Improve error messages for better user experience.

Fixes https://github.com/openshift/origin/issues/9570
Fixes https://github.com/openshift/origin/issues/9585

@liggitt PTAL

### CODE REVIEW

**Comment by liggitt** [comment]
I don't ever want to encourage someone to repeat the `oc login` command against a non-https URL. Also, `oc login` should not be used with the proxy... see https://bugzilla.redhat.com/show_bug.cgi?id=1343008#c3

**Comment by fabianofranz** [comment]
@liggitt Yes, I agree. Added my comment there.

Still, apart the related bug I think these two topics are still worth improving: 1) "forbidden" in the " /" url we ping to check if the server is reachable should not error out ("/" may be forbidden but the server is still accessible, and "/api" may also be accessible); 2) better message when getting that "tls: oversized record" error, I've seen that happen a few times.

So I repurposed this PR to detach from that bug and still address these 2 issues, but I'll leave it to you to decide if it's worth.

File: pkg/cmd/util/clientcmd/errors.go [hunk 1]
```diff
@@ -12,12 +13,15 @@ const (
...
 	configurationInvalidReason        = 3
+	tlsOversizedRecordReason          = 4
 
 	certificateAuthorityUnknownMsg = "The server uses a certificate signed by unknown authority. You may need to use the --certificate-authority flag to provide the path to a certificate file for the certificate authority, or --insecure-skip-tls-verify to bypass the certificate check and use insecure connections."
 	notConfiguredMsg               = `The client is not configured. You need to run the login command in order to create a default config for your server and credentials:
   oc login
 You can also run this command again providing the path to a config file directly, either through the --config flag of the KUBECONFIG environment variable.
 `
+	tlsOversizedRecordMsg          = `Unable to connect to server using TLS, cause: %s.
+If you are using HTTPS to access the server make sure it is supported.`
```
Comment (liggitt) [inline pkg/cmd/util/clientcmd/errors.go]: Ensure the specified server supports HTTPS.

File: test/cmd/authentication.sh [hunk 2]
```diff
@@ -50,8 +50,6 @@ os::cmd::expect_success_and_text 'oc policy can-i create pods --token=${accessto
 os::cmd::expect_success_and_text "oc policy can-i create pods --token=${accesstoken} -n cmd-authentication --scopes='role:admin:*'" 'yes'
 os::cmd::expect_success_and_text "oc policy can-i --list --token=${accesstoken} -n cmd-authentication --scopes='role:admin:*'" 'get.*pods'
 os::cmd::expect_success_and_not_text "oc policy can-i --list --token=${accesstoken} -n cmd-authentication" 'get.*pods'
-
-
```
Comment (liggitt) [inline test/cmd/authentication.sh]: undo empty diff

File: pkg/cmd/cli/cmd/loginoptions.go [hunk 3]
```diff
@@ -159,6 +159,13 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
 				fmt.Fprintln(o.Out)
 			}
 
+		// path "/" we tried to ping is forbidden, but the server is still reachable
+		case clientcmd.IsForbidden(result.Error()):
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/loginoptions.go]: where have we seen this error in a legitimate setup?

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/loginoptions.go]: Why are we trying to ping slash?  I can think of lots of better URLs to ping.

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/loginoptions.go]: @liggitt not that I know. If we decide to ping something else then this will likely not be needed anymore.

@smarterclayton Yeah, it's been here forever but I agree there are much better options. Maybe `/oapi` or `/oauth`?

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/loginoptions.go]: We had osapi before, it broke, then we changed it to /, which broke, so
let's see if we can get it right the third time.

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/loginoptions.go]: @smarterclayton Updated to ping the root of OpenShift API set in `config.APIPath` (`/oapi` at the moment). @liggitt Removed the `IsForbidden` check. PTAL.

Comment (smarterclayton) [inline pkg/cmd/cli/cmd/loginoptions.go]: We will probably remove /oapi at some point in the future.  Use `/apis`,
that is more future proof.  Well.  For now.

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/loginoptions.go]: Ok, updated.

**Comment by liggitt** [comment]
won't `/oapi` fail against a 1.0 server? not sure we granted root access at that point

**Comment by liggitt** [comment]
we should capture the responses of an origin and k8s server at 1.0, 1.1, and 1.2 and have those available as mock servers we can test login and discovery flows against

**Comment by liggitt** [comment]
`/apis` will definitely fail against a 1.0 server

**Comment by smarterclayton** [comment]
So should we check a few urls?  I'd rather find something that matches
as much as possible over being too aggressive on failure.

**Comment by liggitt** [comment]
If all we're doing is checking connectivity for TLS purposes, why do we care about the status code?

File: pkg/cmd/cli/cmd/loginoptions.go [hunk 4]
```diff
@@ -159,6 +159,10 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
 				fmt.Fprintln(o.Out)
 			}
 
+		// TLS oversized record which usually means the server only supports "http"
+		case clientcmd.IsTLSOversizedRecord(result.Error()):
+			return nil, fmt.Errorf(clientcmd.GetPrettyMessageFor(result.Error()))
+
 		default:
 			return nil, result.Error()
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/loginoptions.go]: I'd rather distinguish between connection-level errors and HTTP status code / body content errors... I don't see why we care about a 404 or non-JSON response here, we really just want to make sure the host resolves and TLS connections can be made

**Comment by fabianofranz** [comment]
@liggitt PTAL at the last commit with the proposal for using a TCP connection to `host:port` instead of a full HTTP connection through the client when trying to reach the server first time. I still have to be able to differentiate between a couple errors types to be able to prompt for things like `allow insecure? y/n`, but with this at least we remove the need of an URI (`/`, `/apis`, `/api`, etc).

**Comment by fabianofranz** [comment]
@smarterclayton this also fixes https://github.com/openshift/origin/issues/9570.

**Comment by fabianofranz** [comment]
[test]

**Comment by fabianofranz** [comment]
[test]

**Comment by fabianofranz** [comment]
re[test]

**Comment by fabianofranz** [comment]
[test]

**Comment by fabianofranz** [comment]
@smarterclayton @liggitt more comments?

**Comment by fabianofranz** [comment]
UP

**Comment by smarterclayton** [comment]
I don't have any comments

File: pkg/cmd/cli/cmd/login.go [hunk 5]
```diff
@@ -102,7 +102,7 @@ func (o *LoginOptions) Complete(f *osclientcmd.Factory, cmd *cobra.Command, args
 		o.StartingKubeConfig = kclientcmdapi.NewConfig()
 	}
 
-	addr := flagtypes.Addr{Value: "localhost:8443", DefaultScheme: "https", DefaultPort: 8443, AllowPrefix: true}.Default()
+	addr := flagtypes.Addr{Value: "localhost:8443", DefaultScheme: "https", AllowPrefix: true}.Default()
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login.go]: what does this change do?

File: pkg/cmd/cli/cmd/loginoptions.go [hunk 6]
```diff
@@ -92,17 +98,82 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
...
 		return nil, err
 	}
-	o.Server = serverNormalized
+	o.Config = clientConfig
+
+	return o.Config, nil
+}
+
+func (o *LoginOptions) completeClientConfig(clientConfig *restclient.Config) error {
+	if err := o.normalizeAndDialToServer(); err != nil {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/loginoptions.go]: why mix normalization and dialing? keeping them separate lets us assign clientConfig.Host once, and test normalization separately

File: pkg/cmd/cli/cmd/loginoptions.go [hunk 7]
```diff
@@ -120,57 +190,86 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
...
+		if len(certificateAuthorityData) > 0 {
+			pool := x509.NewCertPool()
+			if !pool.AppendCertsFromPEM([]byte(certificateAuthorityData)) {
+				return fmt.Errorf("could not use the provided CA cert file/data")
+			}
+			tlsConfig.RootCAs = pool
+			rootCAsN = len(pool.Subjects())
+		}
+		glog.V(8).Infof("Dialing TCP connection to %s with TLS and %d root CA's", serverURL.Host, rootCAsN)
+		conn, err := tls.Dial("tcp", serverURL.Host, tlsConfig)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/loginoptions.go]: needs a timeout

File: pkg/cmd/cli/cmd/loginoptions.go [hunk 8]
```diff
@@ -120,57 +190,86 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
...
+		if err != nil {
+			return err
+		}
+		conn.Close()
+	default:
+		glog.V(8).Infof("Dialing TCP connection to %s", serverURL.Host)
+		conn, err := net.Dial("tcp", serverURL.Host)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/loginoptions.go]: needs a timeout

File: pkg/cmd/cli/cmd/loginoptions.go [hunk 9]
```diff
@@ -92,17 +98,82 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
...
+					fmt.Fprintln(o.Out)
+				}
+			// TLS record header errors, like oversized record which usually means the server only supports "http"
+			case tls.RecordHeaderError:
+				return fmt.Errorf(clientcmd.GetPrettyMessageFor(err))
+			default:
+				// try alternative ports before error out
+				if parsedURL, err := url.Parse(o.Server); err == nil {
+					if parsedHost, parsedPort, err := net.SplitHostPort(parsedURL.Host); err == nil {
+						alternativePortFor := map[string]string{"443": "8443"}
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/loginoptions.go]: if I'm reading this right, a user could try to log into https://example.com:443 and we would silently log them into https://example.com:8443? that's a no-no. I could see checking if they didn't specify a port (https://example.com), and suggesting they try https://example.com:8443, but I don't want to automatically send credentials to a server other than what they specified

File: pkg/cmd/util/clientcmd/errors.go [hunk 10]
```diff
@@ -68,6 +81,8 @@ func detectReason(err error) int {
 			return noServerFoundReason
 		case clientcmd.IsConfigurationInvalid(err):
 			return configurationInvalidReason
+		case strings.Contains(err.Error(), "tls: oversized record received"):
+			return tlsOversizedRecordReason
```
Comment (liggitt) [inline pkg/cmd/util/clientcmd/errors.go]: add a unit test that starts up a httptest server, attempts to dial it without TLS, and hits this detection, to make sure we catch any message changes with different go versions

File: pkg/cmd/cli/cmd/loginoptions.go (hunk 6 above)
Comment (liggitt) [inline pkg/cmd/cli/cmd/loginoptions.go]: it makes me a little nervous having this much code spread across multiple functions that loops through stanzas pulling insecure/ca bits out of cluster configs... is there a cleaner way to factor it?

File: pkg/cmd/cli/cmd/login.go (hunk 5 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/login.go]: This is what fixes https://github.com/openshift/origin/issues/9570. With `DefaultPort: 8443`, in case the port wasn't specified it would use `8443` instead of `443`, the default port for the given scheme.

File: pkg/cmd/cli/cmd/loginoptions.go (hunk 7 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/loginoptions.go]: fixed

File: pkg/cmd/cli/cmd/loginoptions.go (hunk 8 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/loginoptions.go]: fixed

File: pkg/cmd/util/clientcmd/errors.go (hunk 10 above)
Comment (fabianofranz) [inline pkg/cmd/util/clientcmd/errors.go]: added

File: pkg/cmd/cli/cmd/loginoptions.go (hunk 9 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/loginoptions.go]: You are right. I changed the approach here, we're only going to suggest the port in `defaultClusterURL` in case that's not what you're already using. But not submitting a request based on assumptions.

**Comment by fabianofranz** [comment]
@liggitt comments addressed, PTAL

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 11]
```diff
@@ -92,75 +97,77 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
 		}
 	}
 
-	// normalize the provided server to a format expected by config
-	serverNormalized, err := config.NormalizeServerURL(o.Server)
-	if err != nil {
+	if err := o.normalizeServerURL(); err != nil {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: I'd prefer to keep the mutations here in the main method... let's keep the old style:
```
serverNormalized, err := config.NormalizeServerURL(o.Server)
if err != nil {
  return nil, err
}
o.Server = serverNormalized
clientConfig.Host = o.Server
```

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 12]
```diff
@@ -92,75 +97,77 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
...
-		}
-	}
-
-	// ping to check if server is reachable
-	osClient, err := client.New(clientConfig)
-	if err != nil {
-		return nil, err
-	}
+	if err := o.dialToServer(); err != nil {
+		clientConfig.Host = o.Server
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: we assign `clientConfig.Host = o.Server` here and in the nil error case... why not do it unconditionally above like we used to?

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 13]
```diff
@@ -170,20 +177,82 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
...
-			ret.Insert(key)
+// dialToServer takes the Server URL normalized and does a TCP dial to make sure the server is reachable
+func (o *LoginOptions) dialToServer() error {
+	serverURL, err := url.Parse(o.Server)
+	if err != nil {
+		return err
+	}
+
+	// load CA data if we have it
+	var certificateAuthorityData string
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: should be []byte

File: pkg/cmd/util/clientcmd/errors.go [hunk 14]
```diff
@@ -1,23 +1,27 @@
...
+	certificateAuthorityUnknownReason
+	configurationInvalidReason
+	tlsOversizedRecordReason
 
 	certificateAuthorityUnknownMsg = "The server uses a certificate signed by unknown authority. You may need to use the --certificate-authority flag to provide the path to a certificate file for the certificate authority, or --insecure-skip-tls-verify to bypass the certificate check and use insecure connections."
 	notConfiguredMsg               = `The client is not configured. You need to run the login command in order to create a default config for your server and credentials:
   oc login
 You can also run this command again providing the path to a config file directly, either through the --config flag of the KUBECONFIG environment variable.
 `
+	tlsOversizedRecordMsg = `Unable to connect to server using TLS, cause: %s.
```
Comment (liggitt) [inline pkg/cmd/util/clientcmd/errors.go]: drop `, cause`, and include the fully resolved server host in the message:

```
Unable to connect to %s using TLS: %s.
Ensure the specified server supports HTTPS.
```

File: pkg/cmd/cli/cmd/login/loginoptions.go (hunk 12 above)
Comment (fabianofranz) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: Right, this was a leftover from when we was doing retries on other ports. Fixed.

**Comment by fabianofranz** [comment]
@liggitt comments addressed, PTAL

File: pkg/cmd/cli/cmd/login/helpers.go [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+}
+
+type existingClientCertificate struct {
+	cAFile string
+	cAData []byte
+}
+
+// findExistingClientCA returns *either* the existing client CA file *or* data for
+// a given host, and true if it exists in the given config
+func findExistingClientCA(host string, kubeconfig clientcmdapi.Config) (*existingClientCertificate, bool) {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: no need for a struct, just return the two values

File: pkg/cmd/cli/cmd/login/helpers.go [hunk 16]
```diff
@@ -0,0 +1,55 @@
...
+	for _, cluster := range kubeconfig.Clusters {
+		if cluster.Server == host {
+			if len(cluster.CertificateAuthority) > 0 {
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: if len(cluster.CertificateAuthority) > 0 || len(cluster.CertificateAuthorityData) > 0 {
  return cluster.CertificateAuthority, cluster.CertificateAuthorityData, true
}

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/login/helpers.go]: I'm ok with removing the struct but I wanted `findExistingClientCA` to only return *one* of `CertificateAuthority` or `CertificateAuthorityData`. Having both in the client config is an error and if somehow both are present in a cluster entry we read here, your suggestion doesn't protect us from that problem. Ok with me keeping these `if`s?

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/login/helpers.go]: Also, if we keep the struct I was thinking about moving [these lines](https://github.com/fabianofranz/origin/blob/0b12ebf90b3d645c172547a2e8c08a82268eae34/pkg/cmd/cli/cmd/login/loginoptions.go#L204-L213) to a `loadFromFile` or similar method in the struct, and remove it from loginoptions.go. What do you think?

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/login/helpers.go]: @liggitt I pushed the changes based on the proposal in these comments, PTAL?

File: pkg/cmd/util/clientcmd/errors.go [hunk 17]
```diff
@@ -1,27 +1,31 @@
...
   oc login
 You can also run this command again providing the path to a config file directly, either through the --config flag of the KUBECONFIG environment variable.
 `
+	tlsOversizedRecordMsg = `Unable to connect to %[2]s using TLS: %[1]s.
+Ensure the specified server supports HTTPS.`
 )
 
 // GetPrettyMessageFor prettifys the message of the provided error
-func GetPrettyMessageFor(err error) string {
+func GetPrettyMessageFor(err error, a ...interface{}) string {
```
Comment (liggitt) [inline pkg/cmd/util/clientcmd/errors.go]: -1 on varargs passed to rangom format strings... if you want a server name, pass it

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 18]
```diff
@@ -102,67 +107,70 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
 	o.Server = serverNormalized
 	clientConfig.Host = o.Server
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: I think the old flow was easier to follow, though it could benefit from some refactoring. I think I'd prefer something like this. Note that we always attempt to pull in the CA first before dialing, dialToServer is stateless, it just uses the passed in clientConfig (and doesn't do CA searching on its own), and that the `o.InsecureTLS` check is only done when the error we encountered was an unknown authority issue (otherwise you could hit a nonexistent server and we'd happily continue on if you also passed `--insecure-skip-tls-verify`)

```
clientConfig := &restclient.Config{}

// load/normalize/prompt for server
clientConfig.Host = ...

// use specified CA or find existing CA
if len(o.CAFile) > 0 {
  clientConfig.CAFile = o.CAFile
  clientConfig.CAData = nil
} else {
  clientConfig.CAFile, clientConfig.CAData = findExistingClientCA(clientConfig.Host, *o.StartingKubeConfig)
}

// attempt to dial using the host/ca data in the client config
if err := dialToServer(clientConfig); err != nil {
  switch err.(type) {
  case x509.UnknownAuthorityError:
    if o.InsecureTLS {
      // set clientConfig to insecure mode
    } else {
      // prompt for insecure mode
    }
  case tls.RecordHeaderError:
    // return error about server not supporting HTTPS
  default:
    h1, p1, err1 := getHostPort(o.Server)
    h2, p2, err2 := getHostPort(defaultClusterURL)
    if err1 == nil && err2 == nil && p1 != p2 {
      // return error, hinting at using different port
    }
    // return error
  }
}
```

**Comment by fabianofranz** [comment]
@liggitt Thanks for the ideas, looking much more readable. The only major change in comparison to what you suggested is that in `dialToServer` we only can make use of CA data. So, if what we have is a file, we need to read it. Mind reviewing?

File: pkg/cmd/cli/cmd/login/helpers.go [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+		if len(caData) > 0 {
+			pool := x509.NewCertPool()
+			if !pool.AppendCertsFromPEM(caData) {
+				return fmt.Errorf("could not use the provided CA cert file/data")
+			}
+			tlsConfig.RootCAs = pool
+			rootCAsN = len(pool.Subjects())
+		}
+		dialer := &net.Dialer{Timeout: timeout}
+		glog.V(8).Infof("Dialing TCP connection to %s with TLS and %d root CA's", serverURL.Host, rootCAsN)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: if we have no CA data, we're using system roots, not 0 root CAs

File: pkg/cmd/cli/cmd/login/helpers.go [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+		conn, err := tls.DialWithDialer(dialer, "tcp", serverURL.Host, tlsConfig)
+		if err != nil {
+			return err
+		}
+		conn.Close()
+	default:
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: make this `case "http":` (we should be normalized, right?), and add a default case that returns an error about an unrecognized scheme

File: pkg/cmd/cli/cmd/login/helpers.go [hunk 21]
```diff
@@ -0,0 +1,139 @@
...
+			pool := x509.NewCertPool()
+			if !pool.AppendCertsFromPEM(caData) {
+				return fmt.Errorf("could not use the provided CA cert file/data")
+			}
+			tlsConfig.RootCAs = pool
+			rootCAsN = len(pool.Subjects())
+		}
+		dialer := &net.Dialer{Timeout: timeout}
+		glog.V(8).Infof("Dialing TCP connection to %s with TLS and %d root CA's", serverURL.Host, rootCAsN)
+		conn, err := tls.DialWithDialer(dialer, "tcp", serverURL.Host, tlsConfig)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: does using this directly mean we lose the proxy behavior of using a transport? seems like a direct dial could fail when a transport+proxy would have succeeded

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/login/helpers.go]: @liggitt Yes, that was my initial concern about not using our own client. But do we really have to care about proxies here since we're only doing TCP? What we care about are web proxies in the HTTP layer so that we can access the REST API.

Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: If a proxy is configured, API requests to myapiserver.com might actual dial myproxy.com, and speak to it instead. Requiring the user to be able to dial myapiserver.com could incorrectly block `oc login` attempts if their network requires them to go through the proxy, and they have no network visibility to myapiserver.com

Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: I think going one layer up and using an http.Transport and calling RoundTrip(req) would be fine

Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: that would also let you do `restclient.TransportFor(&config)` to get the transport, and let it load the CAFile, etc. We probably want to do that anyway, so our debugging hooks wrap the transport and let us see the test dial being done in debug mode

Comment (fabianofranz) [inline pkg/cmd/cli/cmd/login/helpers.go]: Doing a `RoundTrip(req)` with a `HEAD` request is feasible, but that's HTTP already and puts us back to the need of an URL (instead of just host:port)... Which one? 

I'd say we could maybe use a fallback strategy - try simple TCP first, and depending on the error,  fallback to `RoundTrip` with `HEAD` trying a couple places (`/apis`, `/api`, `/`), and only after that fail. Reasonable?

Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: I'm pretty sure we wouldn't care about the RoundTrip response, just the returned error. Request `/`

**Comment by fabianofranz** [comment]
So now to check the server reachability instead of just a TCP `Dial` we're back to HTTP through a `rt.RoundTrip(req)` where `req` is a HEAD request. Our transport is what we get for the existing `restclient.Config` (so with TLS, insecure, proxy, etc, all set - at least all we know up to that point, and we will use what's missing to ask for user input). We discard the response and just use the `error` returned by the round trip as an indication of server reachability.

@liggitt I was finally able to get back to this. PTAL?

**Comment by fabianofranz** [comment]
@liggitt UP

**Comment by liggitt** [comment]
Sorry, on my list for tomorrow

File: pkg/cmd/cli/cmd/login/helpers.go [hunk 22]
```diff
@@ -0,0 +1,117 @@
...
+	matchingClusters := getMatchingClusters(clientConfigToTest, kubeconfig)
+	return len(matchingClusters) > 0
+}
+
+func getHostPort(hostURL string) (string, string, *url.URL, error) {
+	parsedURL, err := url.Parse(hostURL)
+	if err != nil {
+		return "", "", nil, err
+	}
+	host, port, err := net.SplitHostPort(parsedURL.Host)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: SplitHostPort("foo.com") will return an error, since there's no port in the host. Propagating that error means that a failing `oc login https://foo.com` would not suggest `oc login https://foo.com:8443`. If you want the default port to be suggested in that case, tolerate an error returning from SplitHostPort by just returning the original host and an empty port:
```
if host, port, err := net.SplitHostPort(parsedURL.Host); err == nil {
  return host, port, parsedURL, nil
} else {
  return parsedURL.Host, "", parsedURL, nil
}
```

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 23]
```diff
@@ -104,67 +104,45 @@ func (o *LoginOptions) getClientConfig() (*restclient.Config, error) {
...
+		// the server only supports "http"
+		case tls.RecordHeaderError:
+			return nil, clientcmd.GetPrettyErrorForServer(err, o.Server)
 		default:
-			return nil, result.Error()
+			// suggest the port used in the cluster URL by default, in case we're not already using it
+			host, port, parsed, err1 := getHostPort(o.Server)
+			_, defaultClusterPort, _, err2 := getHostPort(defaultClusterURL)
+			if err1 == nil && err2 == nil && port != defaultClusterPort {
+				parsed.Host = fmt.Sprintf("%s:%s", host, defaultClusterPort)
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: net.JoinHostPort(host, defaultClusterPort)

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 24]
```diff
@@ -302,25 +268,12 @@ func (o *LoginOptions) gatherProjectInfo() error {
 		return err
 	}
 
-	projectsList, err := oClient.Projects().List(kapi.ListOptions{})
+	projects, err := oClient.Projects().List(kapi.ListOptions{})
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: undo these changes, they're reverting https://github.com/openshift/origin/pull/10378/files accidentally

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 25]
```diff
@@ -302,25 +268,12 @@ func (o *LoginOptions) gatherProjectInfo() error {
...
-
-	if len(o.DefaultNamespace) > 0 && !projects.Has(o.DefaultNamespace) {
-		// Attempt a direct get of our current project in case it hasn't appeared in the list yet
-		if currentProject, err := oClient.Projects().Get(o.DefaultNamespace); err == nil {
-			// If we get it successfully, add it to the list
-			projectsItems = append(projectsItems, *currentProject)
-			projects.Insert(currentProject.Name)
-		}
-	}
+	projectsItems := projects.Items
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: undo this

File: pkg/cmd/cli/cmd/login/loginoptions.go [hunk 26]
```diff
@@ -337,6 +290,11 @@ func (o *LoginOptions) gatherProjectInfo() error {
 		fmt.Fprintf(o.Out, "Using project %q.\n", o.Project)
 
 	default:
+		projects := sets.String{}
```
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/loginoptions.go]: undo this

File: pkg/cmd/cli/cmd/login/helpers.go (hunk 22 above)
Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: and add a test to make sure `getHostPort("http://foo.com")` returns `"foo.com","",...,nil`

Comment (liggitt) [inline pkg/cmd/cli/cmd/login/helpers.go]: ah, realized we've already normalized the url at this point... this is probably fine as-is, just document the URL must be normalized and have a host:post

File: pkg/cmd/util/clientcmd/errors.go [hunk 27]
```diff
@@ -34,11 +43,27 @@ func GetPrettyMessageFor(err error) string {
...
 }
 
+// GetPrettyErrorFor prettifys the message of the provided error
+func GetPrettyErrorFor(err error) error {
+	return GetPrettyErrorForServer(err, "")
+}
+
+// GetPrettyErrorForServer prettifys the message of the provided error
+func GetPrettyErrorForServer(err error, serverName string) error {
+	return fmt.Errorf(GetPrettyMessageForServer(err, serverName))
```
Comment (liggitt) [inline pkg/cmd/util/clientcmd/errors.go]: `errors.New(GetPrettyMessageForServer(err, serverName))`

**Comment by liggitt** [comment]
just a couple nits and undo the bad rebase bits

**Comment by fabianofranz** [comment]
@liggitt All comments addressed, thanks for catching the bad rebase. Anything else? Otherwise I'll squash.

**Comment by liggitt** [comment]
LGTM

**Comment by fabianofranz** [comment]
[merge]
